File: app/src/main/java/com/einkreader/ui/library/BookListAdapter.java
package com.einkreader.ui.library;

import android.content.Context;
import android.content.SharedPreferences;
import android.graphics.Color;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.einkreader.EInkReaderApp;
import com.einkreader.R;
import com.einkreader.core.storage.BookStorage;

import java.io.File;
import java.util.List;

/**
 * 书籍列表的适配器 —— 把 BookInfo 数据变成列表项显示
 */
public class BookListAdapter extends BaseAdapter {

    private Context context;
    private List<LibraryActivity.BookInfo> books;

    public BookListAdapter(Context context, List<LibraryActivity.BookInfo> books) {
        this.context = context;
        this.books = books;
        preloadProgress();
    }

    private void preloadProgress() {
        if (books == null) return;
        BookStorage storage = EInkReaderApp.getBookStorage();
        if (storage == null) return;
        for (LibraryActivity.BookInfo book : books) {
            if (book.fileKey != null) {
                book.preloadedProgress = storage.loadProgress(book.fileKey);
            }
        }
    }

    @Override
    public int getCount() {
        return books != null ? books.size() : 0;
    }

    @Override
    public Object getItem(int position) {
        return books != null ? books.get(position) : null;
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        if (convertView == null) {
            convertView = LayoutInflater.from(context).inflate(R.layout.item_book, parent, false);
        }

        LibraryActivity.BookInfo book = books.get(position);

        TextView tvTitle = (TextView) convertView.findViewById(R.id.book_title);
        TextView tvInfo = (TextView) convertView.findViewById(R.id.book_info);
        TextView tvProgress = (TextView) convertView.findViewById(R.id.book_progress_text);
        TextView tvCover = (TextView) convertView.findViewById(R.id.book_cover);
        ProgressBar pbProgress = (ProgressBar) convertView.findViewById(R.id.book_progress_bar);

        tvTitle.setText(book.title);

        if (book.dbRecord != null && "epub".equalsIgnoreCase(book.dbRecord.format)) {
            tvCover.setText("EPUB");
            tvCover.setBackgroundColor(0xFF1A548F);
        } else {
            tvCover.setText("TXT");
            tvCover.setBackgroundColor(0xFF333333);
        }

        if (book.dbRecord != null && book.dbRecord.lastReadTime > 0) {
            long t = book.dbRecord.lastReadTime;
            String when = formatRelativeTime(System.currentTimeMillis() - t);
            tvInfo.setText("最近阅读 " + when + "  ·  " + book.info);
        } else {
            tvInfo.setText(book.info);
        }

        int lastChapter = -1;
        int lastPage = -1;
        int totalCh = 0;

        BookStorage.BookProgress prog = book.preloadedProgress;
        if (prog != null) {
            lastChapter = prog.chapterIndex;
            lastPage = prog.pageIndex;
            totalCh = prog.totalChapters;
        }

        if (lastChapter < 0) {
            SharedPreferences prefs = context.getSharedPreferences("eink_reader_prefs", Context.MODE_PRIVATE);
            File f = book.file;
            String fileKey = (f != null) ? f.getName() + "_" + f.length() + "_" + f.lastModified() : "";
            lastChapter = prefs.getInt("lc_" + fileKey, -1);
            lastPage = prefs.getInt("lp_" + fileKey, -1);
            totalCh = prefs.getInt("total_ch_" + fileKey, 0);
            if (lastChapter < 0 && f != null) {
                String filePath = f.getAbsolutePath();
                lastChapter = prefs.getInt("lc_" + filePath, -1);
                lastPage = prefs.getInt("lp_" + filePath, -1);
                totalCh = prefs.getInt("total_ch_" + filePath, 0);
            }
        }

        if (lastChapter >= 0 && totalCh > 0) {
            float pct = (totalCh > 0) ? (lastChapter * 100f / totalCh) : 0;
            pbProgress.setProgress((int) pct);
            pbProgress.setVisibility(View.VISIBLE);
            tvProgress.setText("已读 " + (int)pct + "%  ·  第" + (lastChapter + 1) + "/" + totalCh + "章");
            tvProgress.setVisibility(View.VISIBLE);
        } else if (lastChapter >= 0) {
            pbProgress.setProgress(0);
            pbProgress.setVisibility(View.GONE);
            tvProgress.setText("阅读进度：第" + (lastChapter + 1) + "章");
            tvProgress.setVisibility(View.VISIBLE);
        } else {
            pbProgress.setVisibility(View.GONE);
            tvProgress.setVisibility(View.GONE);
        }

        return convertView;
    }

    /**
     * 把毫秒差值格式化成可读的相对时间
     */
    private static String formatRelativeTime(long deltaMs) {
        if (deltaMs < 0) deltaMs = 0;
        long sec = deltaMs / 1000;
        if (sec < 60) return "刚刚";
        long min = sec / 60;
        if (min < 60) return min + "分钟前";
        long hour = min / 60;
        if (hour < 24) return hour + "小时前";
        long day = hour / 24;
        if (day < 30) return day + "天前";
        long month = day / 30;
        if (month < 12) return month + "个月前";
        return (day / 365) + "年前";
    }
}

File: app/src/main/java/com/einkreader/ui/library/BookListAdapter.kt
package com.einkreader.ui.library

import android.content.Context
import android.content.SharedPreferences
import android.graphics.BitmapFactory
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.TextView
import com.einkreader.EInkReaderApp
import com.einkreader.di.ServiceLocator
import com.einkreader.R
import java.io.File as JavaFile

class BookListAdapter(
    context: Context,
    private var books: List<LibraryActivity.BookInfo>? = null,
    private val nightMode: Boolean = false
) : BaseAdapter() {

    private val ctx = context.applicationContext
    private val prefs: SharedPreferences = ctx.getSharedPreferences("eink_reader_prefs", Context.MODE_PRIVATE)

    /** 封面位图内存缓存（按封面文件路径），避免滚动时主线程重复解码 */
    private val coverBitmapCache = HashMap<String, android.graphics.Bitmap?>()

    init { preloadProgress() }

    private fun preloadProgress() {
        val bl = books ?: return
        val st = ServiceLocator.getBookStorage() ?: return
        for (b in bl) b.preloadedProgress = st.loadProgress(checkNotNull(b.fileKey))
    }

    override fun getCount(): Int = books?.size ?: 0
    override fun getItem(position: Int): Any? = books?.getOrNull(position)
    override fun getItemId(position: Int): Long = position.toLong()

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val view = convertView ?: LayoutInflater.from(ctx).inflate(R.layout.item_book, parent, false)
        val book = books?.get(position) ?: return view
        val rec = book.dbRecord

        // 夜间模式：列表项背景 + 文字
        if (nightMode) {
            view.setBackgroundColor(-0xddddde)
            view.findViewById<TextView>(R.id.book_title).setTextColor(-0x444445)
            view.findViewById<TextView>(R.id.book_info).setTextColor(-0x555556)
            view.findViewById<TextView>(R.id.book_progress_text).setTextColor(-0x444445)
        }

        view.findViewById<TextView>(R.id.book_title).text = book.title

        val tvCover = view.findViewById<TextView>(R.id.book_cover)
        val ivCover = view.findViewById<ImageView>(R.id.book_cover_img)
        // ✅ [Phase 7] format 为 null 时从文件路径推断
        val isEpub = (rec != null && "epub".equals(rec.format, ignoreCase = true)) ||
                     (rec?.filePath != null && checkNotNull(rec.filePath).lowercase().endsWith(".epub"))
        // 优先显示缓存封面图（ReaderActivity 打开书时写入 cacheDir/covers/<key>.jpg）
        val coverKey = checkNotNull(book.fileKey).hashCode().toString(16) + ".jpg"
        val coverPath = JavaFile(ctx.cacheDir, "covers/$coverKey").absolutePath
        val coverBmp = coverBitmapCache.getOrPut(coverPath) {
            val f = JavaFile(coverPath)
            if (f.exists()) {
                try { BitmapFactory.decodeFile(coverPath) } catch (e: Exception) { null }
            } else null
        }
        if (coverBmp != null) {
            ivCover.setImageBitmap(coverBmp)
            ivCover.visibility = View.VISIBLE
            tvCover.visibility = View.GONE
        } else {
            ivCover.visibility = View.GONE
            tvCover.visibility = View.VISIBLE
            if (isEpub) {
                tvCover.text = "EPUB"
                tvCover.setBackgroundColor(-0x46ab71.toInt())
            } else {
                tvCover.text = "TXT"
                tvCover.setBackgroundColor(-0xcccccc.toInt())
            }
        }

        val tvInfo = view.findViewById<TextView>(R.id.book_info)
        if (rec != null && rec.lastReadTime > 0L) {
            tvInfo.text = "最近阅读 ${formatMs(System.currentTimeMillis() - rec.lastReadTime)} · ${book.info}"
        } else {
            tvInfo.text = book.info
        }

        var ch = -1
        var tch = 0
        book.preloadedProgress?.let { ch = it.chapterIndex; tch = it.totalChapters }
        if (ch < 0) {
            val fileObj = book.file ?: return@getView view
            val fkey = getNameKey(fileObj)
            ch = prefs.getInt("lc_$fkey", -1)
            val tc = prefs.getInt("total_ch_$fkey", 0)
            if (tc > 0) tch = tc
            if (ch < 0) {
                val fp = fileObj.absolutePath
                ch = prefs.getInt("lc_$fp", -1)
                val tc2 = prefs.getInt("total_ch_$fp", 0)
                if (tc2 > 0) tch = tc2
            }
        }

        val pb = view.findViewById<ProgressBar>(R.id.book_progress_bar)
        val tp = view.findViewById<TextView>(R.id.book_progress_text)

        if (ch >= 0 && tch > 0) {
            val pct = (ch * 100f / tch).toInt()
            pb.progress = pct
            pb.visibility = View.VISIBLE
            tp.text = "已读 $pct% · 第${ch + 1}章/$tch 章"
            tp.visibility = View.VISIBLE
        } else if (ch >= 0) {
            pb.progress = 0
            pb.visibility = View.GONE
            tp.text = "阅读进度：第${ch + 1}章"
            tp.visibility = View.VISIBLE
        } else {
            pb.visibility = View.GONE
            tp.visibility = View.GONE
        }

        return view
    }

    companion object {
        fun formatMs(deltaMs: Long): String {
            var d = deltaMs
            if (d < 0) d = 0
            val sec = d / 1000
            if (sec < 60) return "刚刚"
            val min = sec / 60
            if (min < 60) return "${min.toString()}分钟前"
            val hour = min / 60
            if (hour < 24) return "${hour.toString()}小时前"
            val day = hour / 24
            if (day < 30) return "${day.toString()}天前"
            val month = day / 30
            if (month < 12) return "${month.toString()}个月前"
            return "${(day / 365).toString()}年前"
        }
    }
}

private fun getNameKey(file: JavaFile): String = file.name + "_" + file.length().toString() + "_" + file.lastModified().toString()
